module/VuFindAdmin/src/VuFindAdmin/Controller/MaintenanceController.php

```diff
@@ -50,10 +50,62 @@ public function homeAction()
         $view = $this->createViewModel();
         $view->caches = $this->serviceLocator->get(\VuFind\Cache\Manager::class)
             ->getCacheList();
+        $view->scripts = $this->getScripts();
         $view->setTemplate('admin/maintenance/home');
         return $view;
     }
 
+    /**
+     * Get a list of the names of scripts available to run thorugh the admin panel.
+     *
+     * @return array
+     */
+    protected function getScripts(): array
+    {
+        // Load the AdminScripts.ini settings
+        $config = $this->serviceLocator->get(\VuFind\Config\PluginManager::class)
+            ->get('AdminScripts')->toArray();
+
+        // Filter out any commands that the current user does not have permission to run:
+        $permission = $this->permission();
+        $filter = function ($script) use ($permission) {
+            return !isset($script['permission'])
+                || $permission->isAuthorized($script['permission']);
+        };
+        return array_filter($config, $filter);
+    }
+
+    /**
+     * Run script action.
+     *
+     * @return mixed
+     */
+    public function scriptAction()
+    {
+        $script = $this->params()->fromRoute('name');
+        $scripts = $this->getScripts();
+        $details = $scripts[$script] ?? null;
+        if (empty($details['command'])) {
+            $this->flashMessenger()->addMessage('Unknown command: ' . $script, 'error');
+        } else {
+            $code = $output = null;
+            exec($details['command'], $output, $code);
+            $successCode = intval($details['successCode'] ?? 0);
+            if ($code !== $successCode) {
+                $this->flashMessenger()->addMessage(
+                    "Command failed; expected $successCode but received $code",
+                    'error'
+                );
+            } else {
+                $this->flashMessenger()->addMessage(
```

Comment:
I'd prefer addSuccessMessage (and addErrorMessage), but this is the common way..

Reply:
I always forget about those options, but I agree that they're more concise. Fixed! I'll open a separate PR to fix the other flash messenger calls in this controller.